core/src/test/java/kafka/test/junit/ClusterTestExtensions.java

```diff
@@ -190,10 +192,10 @@ private void processClusterTest(ExtensionContext context, ClusterTest annot, Clu
 
         ClusterConfig config = builder.build();
```

Comment:
Could you please refactor the code to make `ClusterConfig` immutable? It seems be a chaos that we pass mutable objects everywhere ...

Reply:
Sure, but that could require some efforts here, there are plenty of places directly invoke `ClusterConfig#serverProperties` and add server properties before cluster start. e.g. `KafkaServerKRaftRegistrationTest`.

Reply:
There are places like https://github.com/apache/kafka/blob/0b4e9afee2ace7edf6ff8690e070100b98627836/core/src/test/scala/integration/kafka/server/KafkaServerKRaftRegistrationTest.scala#L74 
need to add extra properties and then restart cluster. If we make ClusterConfig immutable, this may requires more effort to think about how do we handle this scenario. What I want to say is the work could be huge, and overwhelm what we want to address in this JIRA. i.e. define per broker/controller property

Reply:
> Sure, but that could require some efforts here, there are plenty of places directly invoke ClusterConfig#serverProperties and add server properties before cluster start. e.g. KafkaServerKRaftRegistrationTest.

yep, but it is worth the effort. We adopt the builder pattern already, so the built object should be immutable. If the refactor could includes huge changes, we can have a separate PR for that. Or we can refactor them one by one.

1. `ClusterConfig`
2. `BrokerNode`
3. `ControllerNode`

Reply:
I have filed a jira https://issues.apache.org/jira/browse/KAFKA-16560

Reply:
Thanks [Chia-Ping Tsai] ! In this pr, currently I'm working on making ClusterConfig immutable. And other places as you mentioned like `BrokerNode`, `ControllerNode` are not immutable. We can continue the work in the JIRA you just filed.

Reply:
It seems this PR includes the changes to modify `ControllerNode` directly, and that is way I feel the refactor should be completed first.

Reply:
OK, I'll also modify the `ControllerNode` part, thanks !

Reply:
> OK, I'll also modify the ControllerNode part, thanks !

Do you want to do refactor in this PR? or you can complete the refactor in https://issues.apache.org/jira/browse/KAFKA-16560 first. If you prefer to address them at once, I'm ok to close https://issues.apache.org/jira/browse/KAFKA-16560

Reply:
Prefer doing the refactoring part in https://issues.apache.org/jira/browse/KAFKA-16560 since that requires bunch of works to do.

Reply:
I'll address other comments you mentioned excepts the refactoring part tonight. Thanks for your patience.